### PR TITLE
Fixed bugs in OptionalColumnTransform and added bool support

### PR DESCRIPTION
**Pull request by harishsk**
OptionalColumnTransform didn't have support for boolean initializers. When I went to add that I saw that the TensorProto data was being set incorrectly for a few other data types. I tried to fix that and I had to redo the AddInitializer functionality in OnnxContext and related plumbing in OnnxUtils.cs.
Then when I went to add tests for all the data types that are now supported in OptionalColumnTransform, I realized I needed to simplify the column comparison in OnnxConversionTest. In the process I also added support SByte and Byte in OnnxUtils - CreateNamedValue and fixed up the corresponding tests.

In summary:
- Cleaned up OnnxContext's initializer interface
- Cleaned up column comparison functionality on OnnxConversionTest
- Fixed bugs in OptionalColumnTransform's onnx export and added support for boolean initializers

### CODE REVIEW

File: src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs [hunk 1]
```diff
@@ -139,16 +149,17 @@ public OnnxNode CreateNode(string opType, string input, string output, string na
         public abstract string AddInitializer(float value, string name = null, bool makeUniqueName = true);
 
         /// <summary>
-        /// Call this function can declare a global long
+        /// Call this function to declare a global float
```
Comment (ganik) [inline src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs]: >float [](start = 51, length = 5)

long?

File: src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs [hunk 2]
```diff
@@ -157,43 +168,103 @@ public OnnxNode CreateNode(string opType, string input, string output, string na
...
-        /// <param name="dims">The shape that the doubles</param>
+        /// <param name="dims">The shape of values</param>
         /// <param name="name">A string used as a seed to generate this initializer's name in the ONNX graph.</param>
         /// <param name="makeUniqueName">Whether a unique name should be picked for this initializer.</param>
         /// <returns>The initializer's ONNX name</returns>
         public abstract string AddInitializer(IEnumerable<double> values, IEnumerable<long> dims, string name = null, bool makeUniqueName = true);
 
         /// <summary>
-        /// Call this function can declare a global string tensor
+        /// Call this function to declare a global double tensor
```
Comment (ganik) [inline src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs]: >double [](start = 51, length = 6)

ulong?

File: src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs [hunk 3]
```diff
@@ -157,43 +168,103 @@ public OnnxNode CreateNode(string opType, string input, string output, string na
...
-        /// <param name="dims">The shape that the floats</param>
+        /// <param name="dims">The shape of values</param>
         /// <param name="name">A string used as a seed to generate this initializer's name in the ONNX graph.</param>
         /// <param name="makeUniqueName">Whether a unique name should be picked for this initializer.</param>
         /// <returns>The initializer's ONNX name</returns>
         public abstract string AddInitializer(IEnumerable<float> values, IEnumerable<long> dims, string name = null, bool makeUniqueName = true);
 
         /// <summary>
-        /// Call this function can declare a global long tensor
+        /// Call this function to declare a global long tensor
```
Comment (Lynx1820) [inline src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs]: This conflicts with the type param description: "Use this for adding array initializers of integer types smaller than Int32"

File: src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs [hunk 4]
```diff
@@ -139,16 +149,17 @@ public OnnxNode CreateNode(string opType, string input, string output, string na
         public abstract string AddInitializer(float value, string name = null, bool makeUniqueName = true);
 
         /// <summary>
-        /// Call this function can declare a global long
+        /// Call this function to declare a global float
         /// </summary>
-        /// <param name="value">The long number which is going to be added into the ONNX graph</param>
+        /// <param name="value">The float number which is going to be added</param>
+        /// <param name="type">The type of integer to be added, e.g. typeof(short). Use this for all integer types smaller than Int32</param>
```
Comment (Lynx1820) [inline src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs]: Is the type inclusive of Int32s?

File: src/Microsoft.ML.OnnxConverter/OnnxContextImpl.cs [hunk 5]
```diff
@@ -279,13 +279,27 @@ public override List<long> RetrieveShapeOrNull(string variableName)
         }
 
         /// Adds constant tensor into the graph.
+        public override string AddInitializer(bool value, string name = null, bool makeUniqueName = true)
+        {
+            name = AddVariable(name ?? "bool", makeUniqueName);
+            _initializers.Add(OnnxUtils.MakeInt32(name, typeof(bool), value ? 1 : 0));
```
Comment (Lynx1820) [inline src/Microsoft.ML.OnnxConverter/OnnxContextImpl.cs]: Why do booleans get converted to ints?

Comment (harishsk) [reply src/Microsoft.ML.OnnxConverter/OnnxContextImpl.cs]: Please see OnnxML.cs line 2868. According to the protobuf definition, boolean values are stored in the int32Data_ field. 

---
In reply to: [377273444](https://github.com/dotnet/machinelearning/pull/4815#discussion_r377273444) [](ancestors = 377273444)
